trade-finance-manager-ui/server/controllers/login/loginService.js

```diff
@@ -0,0 +1,72 @@
+const { SSO } = require("../../constants");
+
+/**
+ * verifyReferrerForExternalSsoPost
+ * Verify referrer, allow empty referrer for localhost.
+ * 1) If host is localhost then referrer will be missing, no need to check it
+ * 2) Return error message if referrer is missing or it is not from SSO authority
+ * @param {Object} req: Express request
+ * @returns {Boolean|String} False or error message.
+ */
+const verifyReferrerForExternalSsoPost = (req) => {
+  const hostnameWithoutPort = req.get('host').split(':')[0];
+  const referrer = req.get('Referrer');
+
+  // Referrer is not available in localhost because of policy "no-referrer-when-downgrade".
+  if (hostnameWithoutPort === 'localhost') {
+    return false;
+  }
+
+  if (referrer && referrer.indexOf(`${SSO.AUTHORITY}/`) === 0) {
+    return false;
+  }
+
+  console.error('Login request comming from unexpected website: %s', referrer);
+  return 'Login request comming from unexpected website.';
+}
+
+/**
+ * verifyBodyForExternalSsoPost
+ * Verify SSO input parameters, allow limited/safe number of characters.
+ * @param {Object} req: Express request
+ * @returns {Boolean|String} False or error message.
+ */
+const verifyBodyForExternalSsoPost = (templateParams) => {
+  const paramsWithNotAllowedCharacters = Object.entries(templateParams).filter(([, value]) => (/^[0-9a-zA-Z-_.]*$/.test(value) === false));
+
+  if (paramsWithNotAllowedCharacters.length) {
+    console.error('Login request data contains unexpected characters in: %O', paramsWithNotAllowedCharacters);
+    return 'Login request data contains unexpected characters.';
+  }
+
+  return false;
+}
+
+/**
+ * acceptExternalSsoPost
+ * Accept SSO POST from SSO Authority, build form that will auto submit to have session cookie.
+ * 1) Verify referrer.
+ * 2) Verify Post body to have safe parameters.
+ * 3) Return page with form acceptExternalSsoPostForm that will autosubmit.
+ * @param {Object} req: Express request
+ * @param {Object} res: Express response
+ * @returns {Object} response with rendered sso/accept-external-sso-post.njk.
+ */
+const acceptExternalSsoPost = (req, res) => {
```

Comment:
nvm -- its for a js loading spinner i believe

Reply:
Step 1 - we `acceptExternalSsoPost` that is POST from SSO Authority login.microsoftonline.com. Because it is external we don't have TFM session. In sso/accept-external-sso-post.njk we do second, auto submit using `clientside` js.
Step 2 - after internal auto submit we have TFM session and do `processSsoRedirect`.